### DISCUSSION TITLE
Self-paced reading stimuli (static and move window)

### DISCUSSION THREAD

**Discussion by gustavoestivalet** [jspsych/jsPsych#857] [Q&A] [answered]
Hello Guys,

I have followed MovingResearchOnline workshops and learnt a lot, thank you so much! In my lab, we use a lot of self-paced reading experiments in static or move window modalities. I achieve to do a simple lexical decision task, however, when adapting to a self-paced-reading, I have no success at all putting the "test_stimuli" variable in the loop. therefore I have two questions:

1) In the first video of the MRO, it was shown a kind of a self-paced reading task with a single stimulus using a for loop. Then, I tried to put a variable in the for loop to have an experiment with multiple stimuli (many phrases), but no success. Please, could figure out how to organize the "test_stimuli" and loop a variable to present self-paced stimuli (many phrases) word-by-word static in the center of the screen?

2) The next step would be to do a self-paced reading experiment with move window. Could you figure out how to put a move window experiment?

Thank you so much for all your efforts and very nice program!
Best,
Gustavo Estivalet

**Comment by becky-gilbert** [comment, marked as answer]
Hi @gustavoestivalet!

@jodeleeuw has just recently created a new [YouTube tutorial](https://youtu.be/XQcsFwAmbiw) about creating new plugins, and he happens to demonstrate this by creating a moving window self-paced reading plugin! So you could follow along with the video and hopefully that will answer your second question. But if that doesn't do exactly what you need, then let us know.

For your first question about how to organize your test stimuli, you'll want to make an array that stores the information that is specific to each trial, along with any other data you want to save with the trial. So if you have two trials, then it might look something like this:
```javascript
var test_stimuli = [
  {
    words: "This is sentence 1",
    condition: 1
  },
  {
    words: "This is sentence 2",
    condition: 2
  }
];
```
Then you could either construct your trial sequence by looping through your test_stimuli array, constructing a trial using the information from the current object in your array, and then adding that trial on to the timeline:
```javascript
var timeline = [];
for (var i=0; i<test_stimuli.length; i++) {
  var trial = {
    type: 'moving-window',
    words: test_stimuli[i].words,
    data: test_stimuli[i].condition
  };
  timeline.push(trial);
}
```
Or, instead of constructing your trials in a loop, you could set up your trials using [jsPsych's timeline variables](https://www.jspsych.org/overview/timeline/#timeline-variables). 

I hope that helps. Feel free to add your code here if you're still stuck.

**Reply by gustavoestivalet** [reply]
Hi @becky-gilbert , thank you too so much for the fast and exact response! I already saw the videos, and everything is clear for the spr task with mw! Also, the code above seems perfect for looping many trial, I will let you know when I implement it next week with a student!

Enjoying the opportunity, I have a simple question:
Imagine that in the lexical decision experiment I have a tag "fixation" for fixation point and "target" for the targets. Thus, how can I only show/store the "target" information in the end of my experiment (and consequently only save the information related to "target" in MySql)?

Congratulation fo al your nice work!

**Comment by jkhartshorne** [comment]
@gustavoestivalet BTW we'll probably be adding a self-paced reading experiment template to Pushkin in the near future -- with any luck, before the workshop. Pushkin uses jsPsych for stimuli & responses, so that'll mean we'll be making a jsPsych plugin for this purpose, though it may not immediately go into the official jsPsych release. So stay tuned. 

Or, if you have progress making your own plugin, please let us know! This is an open-source project, so we depend on community contributions!

**Reply by gustavoestivalet** [reply]
@jodeleeuw  , thank you so much for all your attention and response. Everything that I needed was in the 3rd session from MRO and the other video in the jsPsych. I have to confess that I am really falling in love by this tool! From the last video, I am just afraid how easy it will be to separate the array of RTs from each segment in different columns to analyze just the segment of interest.

I have not tried yet to implement any plugin. But just for fun, I put below two artesanal solutions that I had before this QA. Let me know what you think!
Solution 1:
`// create test_stimuli
var test_stimuli = [
{fixation: "+", word1: "João", word2: "ama", word3: "Maria.", question: "Paulo ama Maria?", data: {correct_response: "q", type: "no"}},
{fixation: "+", word1: "Carlos", word2: "acorda", word3: "cedo.", question: "Maria acorda cedo?", data: {correct_response: "q", type: "no"}},
{fixation: "+", word1: "Eu", word2: "estou", word3: "cansado.", question: "Eu estou cansado?", data: {correct_response: "p", type: "yes"}},
{fixation: "+", word1: "Ele", word2: "joga", word3: "tênis.", question: "Ele joga tênis?", data: {correct_response: "p", type: "yes"}}
];

// create fixation
var test = {
type: "html-keyboard-response",
stimulus: jsPsych.timelineVariable("fixation"),
choices: "l"
};

// create word1
var test = {
type: "html-keyboard-response",
stimulus: jsPsych.timelineVariable("word1"),
choices: "l"
};

// create word2
var test = {
type: "html-keyboard-response",
stimulus: jsPsych.timelineVariable("word2"),
choices: "l"
};

// create word3
var test = {
type: "html-keyboard-response",
stimulus: jsPsych.timelineVariable("word3"),
choices: "l"
};

// create question
var test = {
type: "html-keyboard-response",
stimulus: jsPsych.timelineVariable("question"),
choices: ["p", "q"],
data: jsPsych.timelineVariable("data"),
on_finish: function(data){
data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
}
};

// create test_procedure
var test_procedure = {
timeline: [fixation, word1, word2, word3, question],
timeline_variables: test_stimuli
};
/ put test_procedure in timeline
timeline.push(test_procedure);
`

Solution 2:
`// create test_stimuli
var test_stimuli = [
{fixation: '+ <span style="color:white">João ama Maria.</span>', word1: '<span style="color:white">+</span> João <span style="color:white">ama Maria.</span>', word2: '<span style="color:white">+ João</span> ama <span style="color:white">Maria.</span>', word3: '<span style="color:white">+ João ama</span> Maria.', question: "Paulo ama Maria?", data: {correct_response: "q", type: "no"}},
{fixation: '+ <span style="color:white">Carlos acorda cedo.</span>', word1: '<span style="color:white">+</span> Carlos <span style="color:white">acorda cedo.</span>', word2: '<span style="color:white">+ Carlos</span> acorda <span style="color:white">cedo.</span>', word3: '<span style="color:white">+ Carlos acorda</span> cedo.', question: "Maria acorda cedo?", data: {correct_response: "q", type: "no"}},
{fixation: '+ <span style="color:white">Eu estou cansado.</span>', word1: '<span style="color:white">+</span> Eu <span style="color:white">estou cansado.</span>', word2: '<span style="color:white">+ Eu</span> estou <span style="color:white">cansado.</span>', word3: '<span style="color:white">+ Eu estou </span> cansado.', question: "Eu estou cansado?", data: {correct_response: "p", type: "yes"}},
{fixation: '+ <span style="color:white">Ele joga tênis.</span>', word1: '<span style="color:white">+</span> Ele <span style="color:white">joga tênis.</span>', word2: '<span style="color:white">+ Ele</span> joga <span style="color:white">tênis.</span>', word3: '<span style="color:white">+ Ele joga</span> tênis.', question: "Ele joga tênis?", data: {correct_response: "p", type: "yes"}}
];`

Sorry if the Solution 2 is not clear, but you can probably figure out that I tried to play with colors!
BTW, where exactly in the CSS file should I change the: a) font, b) font size, c) font color, and d) background color?

**Comment by gustavoestivalet** [comment]
Hi @jodeleeuw and @becky-gilbert , I finally implemented the SPR MW experiment as in the tutorial and it worked like a charm! Thank you so much!

Now, I have some specific issues. In many experiments in my lab we do not present word-by-word exactly, but phrases from a sentence, such as ["The dog", "chased", "the car"]. Therefore, I believe it will be necessary to explicitly define the phrases from the sentence in an array stimuli, no problem, I can follow Becky advice above.

It seems I could just delete the "split" story and pass directly to array (words = ["The dog", "chased", "the car"]) to word_list. However, when the rest of the sentence is reconstructed by hyphens "-", jsPsych also put hyphens in spaces. Thus, how could I keep the spaces in the phrases?

**Deleting split story**
The dog ------ -------
------- chased -------
------- ------ the car

**Final goal!**
The dog ------ --- ---
--- --- chased --- ---
--- --- ------ the car

Thank you so much for all your assistance! All else is working great!

**Reply by jodeleeuw** [reply]
@gustavoestivalet this is helpful feedback for putting together a standard version of this plugin to ship with jsPsych!

There are probably a few ways you could identify spaces. Instead of using the `.repeat()` method that I described in the tutorial you could create a string in a loop:

```js
var output_string = "";
for(var i = 0; i < word.length; i++){
	if(word[i] != " "){
		output_string += "-";
	} else {
		output_string += " ";
	}
}
```

**Reply by gustavoestivalet** [reply]
Hi @jodeleeuw , it just worked perfect! Thank you much, it is done! You are the best! Long life to jsPsych!
Any future project to control LPT1 or USB ports to send triggers to EEG?
Best regards,

**Reply by jodeleeuw** [reply]
> Any future project to control LPT1 or USB ports to send triggers to EEG?

We actually use jsPsych for EEG experiments in my lab. We use a TCP/IP protocol to communicate with our system. Plans are in the works to support Lab Streaming Layer later this year. Feel free to open another discussion if you want to chat more about this. I don't want to bury it in this one :)

**Reply by gustavoestivalet** [reply]
Hi @jodeleeuw  , that sounds very good, I will open a new topic to discuss EEG issues.

Below is my actual code, the presentation of words/phrases are working great, accordingly to Becky instructions!
Question: How/where can I call the "question" object in the end of each trial (sentence)?

`
	// create timeline
	var timeline = [];

	// create stimuli
	var test_stimuli = [
		{words: ["The dog", "chases", "the car."], question: "The cat chases the car?", data: {correct_response: "q"}},
		{words: ["The tiger", "walks", "in the forest."], question: "The tiger walks in the forest?", data: {correct_response: "p"}}
	];

	// create question
	var question = {
		type: "html-keyboard-response",
		stimulus: jsPsych.timelineVariable("question"),
		choices: ["p", "q"],
		data: jsPsych.timelineVariable("data"),
		on_finish: function(data){
		data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
		}
	};

	// show words
	for (var i = 0; i < test_stimuli.length; i++){
		var trial = {
			type: "moving-window-phrase",
			words: test_stimuli[i].words,
		};
		timeline.push(trial);
	}
`

**Reply by gustavoestivalet** [reply]
Hi Guys, I believe I found a nice solution below, does it seems correct?

`// create timeline
var timeline = [];

// create stimuli
var test_stimuli = [
	{words: ["The dog", "chases", "the car."], question: "The cat chases the car?", data: {correct_response: "q"}},
	{words: ["The tiger", "walks", "in the forest."], question: "The tiger walks in the forest?", data: {correct_response: "p"}}];


// create trial
var trial = {
	type: "moving-window-phrase",
	words: jsPsych.timelineVariable("words"),
};

// create question
var question = {
	type: "html-keyboard-response",
	stimulus: jsPsych.timelineVariable("question"),
	choices: ["p", "q"],
	data: jsPsych.timelineVariable("data"),
	on_finish: function(data){
	data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
	}
};

// create test_procedure
var test_procedure = {
	timeline: [trial, question],
	timeline_variables: test_stimuli
};
timeline.push(test_procedure);`

**Reply by becky-gilbert** [reply]
Looks good to me!

**Reply by gustavoestivalet** [reply]
Thank you @becky-gilbert and @jodeleeuw ! It is working great, I tested 3 pilot participants today and the experiment is working perfect in Google Chrome, Mozilla and IE! Thank you so much for all your attention and assistance!

**Reply by becky-gilbert** [reply]
@gustavoestivalet great, glad to hear it's working!
It looks like you had a question about saving the RT values to your database, but then edited your question. Did you find a solution to that problem?

**Reply by gustavoestivalet** [reply]
Hi @becky-gilbert , yes, it is solved. I just changed the type of data of the RT variable in MySql (VARCHAR!), it was INT. Now, the whole array with RTs appears as a string! Thank you so much!

But I do have a question! I have some tasks in Portuguese with diacritics (é, á, ã, etc.). I have had some problems in the the jsPsych display, then I put the line _meta charset="utf-8"_ in the head of the experiment and it solved the display.

Now, I am having character problems in the MySql database. I already set all MySql collation to "utf8mb4_unicode_ci", but no solution. Do you have any hint in how to solve it?

**Reply by gustavoestivalet** [reply]
Hi, Guys, I updated the PHP version in my server to 7.4 and all character/collation errors were corrected. Done!

**Comment by gustavoestivalet** [comment]
Hello guys, thank you so much for all your helping!
Thinking in a mobile experiment, I wonder to figure out how to change the "keyboard press" by a "button press" in this moving window plugin. I already tried to copy code form "jspsych-html-button-response" plugin, but I could not adapt it yet.
Please, if you have in insight in how to use a button instead of a key to pass words/phrases, let me know!
Best,

**Reply by vijaymarupudi** [reply]
I would recommend posting a new discussion issue on this.
